benches/read_tiff.rs

```diff
@@ -0,0 +1,97 @@
+use async_tiff::metadata::{PrefetchBuffer, TiffMetadataReader};
+use async_tiff::reader::ObjectReader;
+use async_tiff::{ImageFileDirectory, Tile};
+use criterion::{criterion_group, criterion_main, Criterion, Throughput};
+use object_store::path::Path;
+use object_store::{parse_url, ObjectStore};
+use rayon::iter::{IntoParallelIterator, ParallelIterator};
+use rayon::ThreadPoolBuilder;
+use reqwest::Url;
+use tokio::runtime;
+
+fn read_tiff(fpath: &str) -> AsyncTiffResult<()> {
+    let abs_path: PathBuf = std::path::Path::new(fpath).canonicalize()?;
+    let tif_url: Url = Url::from_file_path(abs_path).expect("Failed to parse url: {abs_path}");
+    let (store, path): (Box<dyn ObjectStore>, Path) = parse_url(&tif_url)?;
+
+    let reader = ObjectReader::new(Arc::new(store), path);
+    let decoder_registry = DecoderRegistry::default();
+
+    // Initialize async runtime
+    let runtime = runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()?;
+
+    // Get list of tiles in TIFF file stream (using tokio async runtime)
+    let tiles: Vec<Tile> = runtime
+        .block_on(async {
+            // Read metadata header
+            let prefetch_reader = PrefetchBuffer::new(reader.clone(), 32 * 1024).await?;
+            let mut metadata_reader = TiffMetadataReader::try_open(&prefetch_reader).await?;
+
+            // Read Image File Directories
+            let ifds: Vec<ImageFileDirectory> =
+                metadata_reader.read_all_ifds(&prefetch_reader).await?;
+
+            assert_eq!(ifds.len(), 1); // should have only 1 IFD
+            let ifd: &ImageFileDirectory = ifds.first().ok_or(AsyncTiffError::General(
+                "unable to read first IFD".to_string(),
+            ))?;
+
+            let (x_count, y_count) = ifd.tile_count().ok_or(AsyncTiffError::General(
+                "unable to get IFD count".to_string(),
+            ))?;
+            // dbg!(x_count, y_count); // 43 * 43 = 1849
+
+            // Get cartesian product of x and y tile ids
+            let x_ids: Vec<usize> = (0..x_count)
+                .flat_map(|i| (0..y_count).map(move |_j| i))
+                .collect();
+            let y_ids: Vec<usize> = (0..x_count).flat_map(|_i| 0..y_count).collect();
+
+            let tiles: Vec<Tile> = ifd.fetch_tiles(&x_ids, &y_ids, &reader).await?;
+            assert_eq!(tiles.len(), 1849);
+
+            Ok::<Vec<Tile>, AsyncTiffError>(tiles)
+        })
+        .unwrap();
+
+    // Do actual decoding of TIFF tile data (multi-threaded using rayon)
+    let pool = ThreadPoolBuilder::new()
+        .num_threads(4)
+        .build()
+        .map_err(|err| AsyncTiffError::External(Box::new(err)))?;
+
+    let tile_bytes: Vec<u8> = pool.install(|| {
+        tiles
+            .into_par_iter()
+            .flat_map_iter(|tile| tile.decode(&decoder_registry).unwrap())
+            .collect()
+    });
+    assert_eq!(tile_bytes.len(), 363528192); // should be 361681200, why not?
```

Comment:
Hmm what's happening here? Can you explain why you expect the number to be different?

Is it possible there's an issue with whether this is compressed/decompressed? Or whether there's a step in the decoding missing?

Reply:
It should be 10980x10980x3bands = 361681200 bytes (since this is u8 dtype, so 1 byte per pixel). The number is not lower, but higher, so the decompression should have worked, but not sure what's causing the deviation. I should probably try to find a way to piece the bytes together and display it to see what's going on.

Reply:
Maybe read it in python to inspect it?